File: Binary_Tree/tree.py
import collections


class BinaryTree:
    def __init__(self, val = None, left = None, right = None):
        self.val = val
        self.left = left
        self.right = right


def build_tree(array):
    root = None
    if len(array) > 0:
        array.sort()
        def recur(left, right):
            if left > right:
                return None
            mid = (left + right) // 2
            return BinaryTree(
                val = array[mid], 
                left = recur(left, mid - 1), 
                right = recur(mid + 1, right)
            )
        root = recur(0, len(array) - 1)
    return root


def max_height(root):

    def recur(node, depth):

        if node == None:
            return depth

        return max(
            recur(node.left, depth + 1),
            recur(node.right, depth + 1)
        )

    return recur(root, -1)

## @ https://stackoverflow.com/questions/48850446/how-to-print-a-binary-tree-in-as-a-structure-of-nodes-in-python/58104529#58104529
def print_tree(root):
    res = []
    q = collections.deque()
    max_depth = max_height(root)
    q.append((root, 0))
    while q:
        row = []
        for _ in range(len(q)):
            node, depth = q.popleft()
            if not node:
                if depth <= max_depth:
                    row.append("#")
                continue
            row.append(node.val)
            q.append((node.left, depth + 1))
            q.append((node.right, depth + 1))
        res.append(row)
    rows = len(res)
    base = 2**(rows)
    for r in range(rows):
        for v in res[r]:
            print("." * (base), end = "")
            print(v, end = "")
            print("." * (base - 1), end = "")
        print("|")
        base //= 2



# array = [x for x in range(10)]
# root = build_tree(array)
# print_tree(root)

File: Binary_Tree/tree.cpp
#include <vector>
#include <bits/stdc++.h>

struct BinaryTree {
    int val;
    BinaryTree* left;
    BinaryTree* right;
    BinaryTree(): val(0), left(nullptr), right(nullptr) {}
    BinaryTree(int x) : val(x), left(nullptr), right(nullptr) {}
    BinaryTree(int x, BinaryTree* left, BinaryTree* right) : val(x), left(left), right(right) {}
};

BinaryTree* build_tree_recur(int left, int right, std::vector<int> &array) {
    if (left > right) {
        return nullptr;
    }
    int mid = (left + right) / 2;
    return new BinaryTree(
        array[mid],
        build_tree_recur(left, mid - 1, array),
        build_tree_recur(mid + 1, right, array)
    );
}

BinaryTree* build_tree(std::vector<int> &array) {
    BinaryTree* root;
    if (array.size() > 0) {
        std::sort(array.begin(), array.end());
        root = build_tree_recur(0, array.size() - 1, array);
    }
    return root;
}

// @ https://stackoverflow.com/questions/13484943/print-a-binary-tree-in-a-pretty-way
void print_tree(BinaryTree* p, int indent = 1)
{
    if(p != NULL) {
        if(p->right) {
            print_tree(p->right, indent+4);
        }
        if (indent) {
            std::cout << std::setw(indent) << ' ';
        }
        if (p->right) std::cout<<" /\n" << std::setw(indent) << ' ';
        std::cout<< p->val << "\n ";
        if(p->left) {
            std::cout << std::setw(indent) << ' ' <<" \\\n";
            print_tree(p->left, indent+4);
        }
    }
}

// int main() {

//     std::vector<int> array = {4, 1, 5, 6, 2, 6,2, 52,5 ,6 ,15 ,6 ,163 , 3};
//     BinaryTree* root = build_tree(array);
//     postorder(root);
//     return 0;
// }
